src/node_http2.h

```diff
@@ -468,75 +657,369 @@ class Http2Session : public AsyncWrap,
-  int ReadStop() override { return 0; }
+  int DoWrite(WriteWrap* w, uv_buf_t* bufs, size_t count,
+              uv_stream_t* send_handle) override;
 
-  // Required for StreamBase
-  int DoShutdown(ShutdownWrap* req_wrap) override {
-    return 0;
-  }
+  size_t self_size() const override { return sizeof(*this); }
 
-  uv_loop_t* event_loop() const override {
-    return env()->event_loop();
+  // Handling Trailer Headers
+  class SubmitTrailers {
+   public:
+    inline void Submit(nghttp2_nv* trailers, size_t length) const;
+
+    inline SubmitTrailers(Http2Session* sesion,
+                          Http2Stream* stream,
+                          uint32_t* flags);
+
+   private:
+    Http2Session* const session_;
+    Http2Stream* const stream_;
+    uint32_t* const flags_;
+
+    friend class Http2Stream;
+  };
+
+  void OnTrailers(const SubmitTrailers& submit_trailers);
+
+  // JavaScript API
+  static void GetID(const FunctionCallbackInfo<Value>& args);
+  static void Destroy(const FunctionCallbackInfo<Value>& args);
+  static void FlushData(const FunctionCallbackInfo<Value>& args);
+  static void Priority(const FunctionCallbackInfo<Value>& args);
+  static void PushPromise(const FunctionCallbackInfo<Value>& args);
+  static void RefreshState(const FunctionCallbackInfo<Value>& args);
+  static void Info(const FunctionCallbackInfo<Value>& args);
+  static void RespondFD(const FunctionCallbackInfo<Value>& args);
+  static void Respond(const FunctionCallbackInfo<Value>& args);
+  static void RstStream(const FunctionCallbackInfo<Value>& args);
+
+  class Provider;
+
+ private:
+  Http2Session* session_;                       // The Parent HTTP/2 Session
+  int32_t id_;                                  // The Stream Identifier
+  int32_t code_ = NGHTTP2_NO_ERROR;             // The RST_STREAM code (if any)
+  int flags_ = NGHTTP2_STREAM_FLAG_NONE;        // Internal state flags
+
+  uint32_t max_header_pairs_ = DEFAULT_MAX_HEADER_LIST_PAIRS;
+  uint32_t max_header_length_ = DEFAULT_SETTINGS_MAX_HEADER_LIST_SIZE;
+
+  // The Current Headers block... As headers are received for this stream,
+  // they are temporarily stored here until the OnFrameReceived is called
+  // signalling the end of the HEADERS frame
+  nghttp2_headers_category current_headers_category_ = NGHTTP2_HCAT_HEADERS;
+  uint32_t current_headers_length_ = 0;  // total number of octets
+  std::vector<nghttp2_header> current_headers_;
+
+  // Inbound Data... This is the data received via DATA frames for this stream.
+  std::queue<uv_buf_t> data_chunks_;
+
+  // Outbound Data... This is the data written by the JS layer that is
+  // waiting to be written out to the socket.
+  std::queue<nghttp2_stream_write*> queue_;
+  unsigned int queue_index_ = 0;
+  size_t queue_offset_ = 0;
+  int64_t fd_offset_ = 0;
+  int64_t fd_length_ = -1;
+};
+
+class Http2Stream::Provider {
+ public:
+  Provider(Http2Stream* stream, int options);
+  explicit Provider(int options);
+  ~Provider();
```

Comment:
Not making a base class destructor virtual is kind of dangerous… The small overhead might be worth avoiding the surprise of future crashes?

Reply:
Yeah, good point :)